### ISSUE TITLE
Break up examples and servant into separate repo's

### ISSUE THREAD

**Issue by tysonzero** [liqula/react-hs#13]
IMO it would be much better to have the 3 separated like before, is there a reason for putting them all together?

**Comment by tysonzero** [comment]
It would be nice if all the github stuff like issues and pull requests and even just commits were isolated to the relevant libraries. Also decoupling examples is nice to show that they truly are examples and not in any way integrated into or overly couple to the repo iteself.

**Comment by fisx** [comment]
sorry, i think the opposite is true.

i think it's considerably more painful to have 3 repositories that need to be kept in sync.  i don't think this is about modularity: we still need to keep things separate to get the individual packages on hackage, but if they share a repository, we have to manage one third the commits where changes spread across packages.  and users do not have to look around for the right repository to open an issue in.

i've seen in servant that having the core packages put together in one repo works very well, and that it's enough to take out less essential packages as the project gets bigger.

what do other people think?

**Comment by fisx** [comment]
I've just added a stack.yaml snippet to the README.  Even that part is getting easier!  (-:

**Comment by tysonzero** [comment]
I can maybe see it being useful to have the examples repo be together. But react-hs-servant is a fundamentally different package that react-hs should not care about in the slightest. I know I personally have no intention of using it the near future. I just care about react-hs.

**Comment by tysonzero** [comment]
Cluttering the react-hs commit history and issue list with commits for a package that it does not depend on or care about in any way is IMO the wrong choice. I will concede that bundling in examples isn't the worst idea. Perhaps they can just be in an `examples` folder in react-hs itself, as long as we make sure they are set up so they can be cut and pasted out into an unrelated directory and work. But yeah IMO nix react-hs-servant.

**Comment by fisx** [comment]
I'm still not convinced.  Would you say that instead of lens there should be lens, prism, iso, ... packages just because some people have no intention of using prisms, but want lenses?

And that's not even the point.  I agree that react-hs-servant should be a separate package, that has to be installed individually from hackage, but that's still the case here.

The question for me is this: will somebody who has a problem with either react-hs or react-hs-servant have an easy time figuring out where to ask?  If the two packages are sufficently entangled, it's easier to have one place instead of two.

I personlly feel there is no possible reason for not using react-hs-servant, whether you are running servant on the backend or not.  (But don't worry, I have no intention of making it the only option. ;-)

In conclusion: I'm not convinced, but if there are enough users and you get a clear enough majority for your position, I will be happy to give in.

**Comment by divipp** [comment]
As I understand @tysonzero would like to separate the commits and issues list of `react-hs`.

Even if `react-hs`, `react-hs-examples` and `react-hs-servant` is in the same repository, it is easy to separate the commits for `react-hs` by clicking on `history` in the `react-hs/react-hs` directory: https://github.com/liqula/react-hs/commits/master/react-hs

Separating the issues of `react-hs` is possible too if we maintain a `react-hs` flag similar to existing flags like `question`, `easy` and `enhancement`.

@tysonzero: Would this be a solution or at least a workaround for you?

**Comment by fisx** [comment]
i like the idea.  i've added the labels for issues / PRs regarding only the non-core packages, but no such issues exist right now.

(i did find another reaeson to keep the repo together, though: #12 would be two issues, two PRs, and a lot of overhead to keep things in sync during development.  now it's much easier.)

**Comment by tysonzero** [comment]
> Would you say that instead of lens there should be lens, prism, iso, ... packages just because some people have no intention of using prisms, but want lenses?

I mean it depends, perhaps, perhaps not, granularity is of course a tradeoff. But that is a totally different argument, since we are talking about whether or not two things that are already different packages belong in the same github repo. I would say no.

> If the two packages are sufficently entangled

But that's the thing, the packages should not be "entangled" at all, that is the exact thing I am trying to avoid. There should be a one way dependency, and absolutely zero thought given in the other direction. I mean of course if react-hs-servant is having trouble due to some issue with react-hs then the issue should be resolved, but react-hs-servant should be treated identically to any and all other libraries that depend on react-hs.

> (i did find another reaeson to keep the repo together, though: #12 would be two issues, two PRs, and a lot of overhead to keep things in sync during development. now it's much easier.)

But fundamentally the issue does not apply in the same way to both repos, either `React.Flux.Servant` needs to use `React.Flux.Ajax` better to avoid code duplication, and react-hs doesn't need to be involved, or `React.Flux.Ajax` needs to export more stuff, and `React.Flux.Servant` then later on needs to use more of that stuff once there has been a release with it exported. I mean I hope you aren't planning on changing `React.Flux.Servant` to import things that only exist in HEAD of react-hs and not in any published version, so you have to wait some time between solving each of the two separate issues anyway.

**Comment by tysonzero** [comment]
@divipp That helps, I still think fundamentally the repos should be separate, but that does help, thanks.

**Comment by fisx** [comment]
I don't think we're going to come to a consensus here, and I'm worried that if I let you split up the packages into separte repos, somebody would come along and open an issue about merging them back together.  Finally, there are more important issues like #10, #11, and all the others that make a difference for users who pull from hackage, not from github.

So I will close this for now.  If react-hs will take over react-flux and this still bothers you then, please re-open this and we can have a poll.

**Comment by tysonzero** [comment]
@fisx I highly doubt such a merging request will come up, this structuring of multiple separate independent packages being in a single repo is extremely uncommon, and not what people expect.
